### PR TITLE
JS Components: typescriptify Dialog component

### PR DESCRIPTION
**Pull request by retrofox**
This PR rewrites the Dialog component using TypeScript

Fixes #

#### Changes proposed in this Pull Request:

* JS Components: typescriptify Dialog component

#### Testing instructions:

* Use Storybook:
```
cd projects/js-packages/storybook && pnpm run storybook:dev
```

* Take a look at
  * Dialog stories: http://localhost:50240/?path=/story/js-packages-components-dialog--interstitial-jetpack-boost
  * Protect Interstitial: http://localhost:50240/?path=/story/plugins-protect-interstitial--default
  * Protect Footer: http://localhost:50240/?path=/story/plugins-protect-footer--default
* Confirm that all the stories work as expected

_Disclaimer: there are some visual issues that we'll be addressed in follow-up PRs_

image

### CODE REVIEW

File: projects/js-packages/components/components/layout/container/index.jsx [hunk 1]
```diff
@@ -48,6 +48,9 @@ Container.propTypes = {
 	horizontalSpacing: PropTypes.number,
 	/** Number of gap betwen rows, it gets multipled by 8px. Needs to be an integer */
 	horizontalGap: PropTypes.number,
+
+	/** Children to be inserted. */
+	children: PropTypes.node,
```
Comment (retrofox) [inline projects/js-packages/components/components/layout/container/index.jsx:53]: needed to follow one
image

File: projects/js-packages/components/components/dialog/index.tsx [hunk 2]
```diff
@@ -18,9 +24,9 @@ import styles from './style.module.scss';
  * @param {React.Component} props.primary   - Primary content.
  * @param {React.Component} props.secondary - Secondary content.
  * @param {boolean} props.split			    - Split the sections.
- * @returns {object}                          Dialog react component.
+ * @returns {React.FunctionComponent}         Dialog react component.
```
Comment (manzoorwanijk) [inline projects/js-packages/components/components/dialog/index.tsx:27]: `FunctionComponent`  means which can be instantiated i.e. which can be used as a component.
```suggestion
 * @returns {React.ReactNode}         Rendered dialog
```

File: projects/js-packages/components/components/dialog/index.tsx [hunk 3]
```diff
@@ -35,14 +41,10 @@ export default function Dialog( { primary, secondary, split } ) {
...
-
-Dialog.propTypes = {
-	split: PropTypes.bool,
 };
 
 Dialog.defaultProps = {
-	primary: null,
-	secondary: null,
 	split: false,
 };
```
Comment (manzoorwanijk) [inline projects/js-packages/components/components/dialog/index.tsx:48]: I think we can remove `defaultProps` and directly use the default values via ES6 default params.

File: projects/js-packages/components/components/dialog/index.tsx [hunk 4]
```diff
@@ -18,9 +24,9 @@ import styles from './style.module.scss';
  * @param {React.Component} props.primary   - Primary content.
  * @param {React.Component} props.secondary - Secondary content.
```
Comment (renatoagds) [inline projects/js-packages/components/components/dialog/index.tsx:25]: Looks like you need to update here too. You receive `React.ReactNode` here.

File: projects/js-packages/components/components/dialog/stories/index.tsx [hunk 5]
```diff
@@ -56,6 +56,12 @@ InterstitialJetpackBoost.args = {
 			pricing={ {
 				isFree: true,
 			} }
+			className={ '' }
+			icon={ '' }
+			isCard={ false }
+			supportedProducts={ [] }
+			hasRequiredPlan={ false }
+			addProductUrl={ '' }
```
Comment (manzoorwanijk) [inline projects/js-packages/components/components/dialog/stories/index.tsx:64]: These shouldn't be required. If that's the case, we can make those props optional in JS docs of `ProductOffer` component. For example, we can replace:
`@param {string} props.className`
with
`@param {string} [props.className]`

File: projects/js-packages/components/components/dialog/stories/declaration.d.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+declare module '*.module.scss';
```
Comment (manzoorwanijk) [inline projects/js-packages/components/components/dialog/stories/declaration.d.ts:1]: Do mean this instead?
```suggestion
declare module '*.png';
```

File: projects/js-packages/components/components/dialog/stories/index.tsx (hunk 5 above)
Comment (retrofox) [reply projects/js-packages/components/components/dialog/stories/index.tsx:64]: yes, I agree. I didn't want to address it here, though. I suggest doing it in a follow-up one.

**Comment by retrofox** [comment]
I realize I need a crew ✅ . cc @renatoagds

**Review by renatoagds — APPROVED** [review]
LGTM

**Comment by retrofox** [comment]
> One last step: head over to your WordPress.com diff, D80317-code, and deploy it.

r245186-wpcom

**Pull request merged**